packages/tasit-action/src/TasitAction.js

```diff
@@ -19,117 +22,370 @@ class Utils {
+  // How we'll should keeping track of nonces?
+  waitForMessage = async () => {
     const tx = await this.#txPromise;
-    this.#provider.removeAllListeners(tx.hash);
+
+    await this.#provider.waitForTransaction(tx.hash);
   };
 }
 
-class EventSubscription {
-  #eventNames;
+class ContractSubscription extends Subscription {
+  #contract;
   #provider;
+  #eventNames;
 
-  constructor(eventNames, provider) {
-    this.#eventNames = eventNames;
+  // Why eventNames shoud be passed here?
+  constructor(contract, provider, eventNames) {
+    eventNames.forEach(eventName => {
+      if (contract.interface.events[eventName] === undefined)
+        throw new Error(`Event '${eventName}' not found.`);
+    });
+
+    super(contract);
+    this.#contract = contract;
     this.#provider = provider;
+    this.#eventNames = eventNames;
   }
 
-  on = (eventName, callback) => {
-    if (!this.#eventNames.includes(eventName))
+  on = (eventName, listener) => {
+    if (!this.#eventNames.includes(eventName) && eventName !== "error") {
       throw new Error(
         `This subscription isn't subscribed on '${eventName}' event.`
       );
+    }
+
+    if (eventName === "error") {
+      this.addErrorListener(listener);
+      return;
+    }
+
+    const wrappedListener = async (...args) => {
+      const event = args.pop();
+      const message = {
+        data: {
+          args: event.args,
+        },
+      };
+
+      try {
+        await listener(message);
+      } catch (error) {
+        this.emitErrorEvent(
+          new Error(`Callback function with error: ${error.message}`)
+        );
+      }
+    };
+
+    this.addListener(eventName, eventName, listener, wrappedListener);
   };
+}
 
-  //removeAllListeners = async () => {};
+class ProviderFactory {
+  static getProvider = () => {
+    let json;
+    try {
+      json = PROVIDER_CONFIG;
+    } catch (error) {
+      console.warn(
+        `Error on parsing config.json file, using default configuration.`
+      );
+      json = ProviderFactory.getDefaultConfig();
+    }
+
+    return ProviderFactory.createProvider(json);
+  };
+
+  static getDefaultConfig = () => {
+    return {
+      network: "other",
+      provider: "jsonrpc",
+      pollingInterval: 50,
+      jsonRpc: {
+        url: "http://localhost:8545",
+        user: "",
+        password: "",
+        allowInsecure: true,
+      },
+    };
+  };
+
+  static createProvider = ({
+    network,
+    provider,
+    pollingInterval,
+    jsonRpc,
+    infura,
+    etherscan,
+  }) => {
+    const networks = ["mainnet", "rinkeby", "ropsten", "kovan", "other"];
+    const providers = ["default", "infura", "etherscan", "jsonrpc"];
+
+    if (!networks.includes(network)) {
+      throw new Error(`Invalid network, use: [${networks}].`);
+    }
+
+    if (!providers.includes(provider)) {
+      throw new Error(`Invalid provider, use: [${providers}].`);
+    }
+
+    if (network === "mainnet") network = "homestead";
```

Comment:
I've done that because I think that `mainnet` is more friendly than `homestead` and I've added another parse like that:
```javascript
if (provider === "fallback") network = "default";
```
`fallback`, `mixed` or `infuraAndEtherscan` sounds more clear than `default` to me.

But we can keep same EthersJs names if you think that will be better (or do that in a better way). What do you think?

Reply:
`fallback` works for me (unless we want to make them explicitly select the fallback and default not erroring out - I am kind of leaning towards that).

`mainnet` works for me!

Reply:
I've thought about `fallback` word: EthersJS allows the user to create a custom fallback provider using an array of provider. Actually, the `DefaultProvider` is a FallbackProvider with Infura and Ethrescan.
E.g.: Some user would create a FallbackProvider using N geth nodes.
Do you think that will be a use case for TasitSDK?

A further question about provider: We aren't supporting `web3` provider, should we?

https://docs.ethers.io/ethers.js/html/api-providers.html#connecting-to-ethereum

Reply:
Good example. I think that people might want to use N geth.

I don't think customizing fallback is something people will need in the SDK. Let's wait for 3 people to ask for that before building it.

Same goes for web3 provider - let's build it as soon as it becomes clear it's desired.

Reply:
Could I create an issue (`enhancement` and `question`) called "Fallback and Web providers support" with a note that it should be done when these features are desired by users?

Reply:
Yes, that sounds great. Please do that!

Reply:
Done. https://github.com/tasitlabs/TasitSDK/issues/111

Reply:
Thanks. Resolving this.